### PR TITLE
Statements refactor

### PR DESCRIPTION
**Pull request by jhannus**
Here are two separated commits of the statement refactor @pezra and I talked about. The type hinting was only applied to some of the writers. I'm not exactly sure I like the way the type hints are being passed around so maybe I could get some feedback/suggestions on it before continuing?

p.s. some tests may be failing ATM

Started from discussion: https://github.com/cequel/cequel/issues/306

### CODE REVIEW

**Comment by pezra** [comment]
I didn't mean to push your branch, @jhannus. I assumed the push would be blocked due to access control. I hope you don't mind.

**Comment by jhannus** [comment]
No worries at all just possible failing specs lol. I'll take a look this coming morning +1

**Comment by pezra** [comment]
If you don't mind i'll keep chipping away at the specs until they are all passing.

**Comment by pezra** [comment]
I also rebased. In for a penny, in for a pound, as they say.

**Comment by jhannus** [comment]
Yes please! I was hoping to find a better way to pass the type hints to the writers. So that's not finished. 

I'm thinking that instead of passing table and columns names to the writers we can pass a schema and figure out the names from there. Since the writers are responsible for assembling the query... We could then figure the queries column types hints from there. 

With writing to a single encapsulated statement object each query whether single or batch, prepared or not, it will contain exactly it's own parameters.

**Comment by pezra** [comment]
@jhannus, let me know what you think. Type hints turned out be to really hard to get working. Once i switch to just preparing everything it all pretty much fell into place.

**Comment by jhannus** [comment]
👍 @pezra thank you. Yeah the type hints where not fun at all but required to optionally prepare. In the fork I have hosted internally thats the way I went. I figured to contribute back though it should be optional possibly for folks who dynamically generate queries but... I'm cool with all prepared all the time if it cool with you 😄

**Comment by jhannus** [comment]
I cherry picked the `PreparedStatementManager` I was talking about plus setters/getters and specs but its still unused on execute. I would remove the keyspace reference from the `Statement` instance (less object refs) and use `Cequel::Metal::Keyspace#prepared_statement` from `#execute` instead.

File: lib/cequel/metal/statement.rb [hunk 1]
```diff
@@ -10,16 +10,30 @@ module Metal
...
-      def cql
-        @cql.join
+      def to_s
+        @cql
+      end
+      alias_method :cql, :to_s
+
+      # @return [Cassandra::Statements::Prepared] prepared version of this statement
+      def prepare(keyspace)
+        @prepared ||= keyspace.client.prepare(cql)
```
Comment (jhannus) [inline lib/cequel/metal/statement.rb]: I think my previous comment on this was canceled somehow but...
Memoizing the prepared CQL is good here but each `Cequel::Metal::Statement` object will ask the server for  a new prepared request even if this statement was already prepared once.

To mitigate this I cherry-picked my `PreparedStatementManager` which reuses prepared statements per `Cequel::Metal::Statement` instance. Right now the `Keyspace` only has the setter/getter for it. We'll need to use `Keyspace#prepare_statement` this during execute instead and remove `Cequel::Metal::Statement#prepare`

**Comment by pezra** [comment]
@jhannus, it looks to me like the cassandra driver caches prepared statements and prevents duplicates (search <https://github.com/datastax/ruby-driver/blob/v2.1.7/lib/cassandra/cluster/client.rb> for `@prepared_statements[host][cql]`). Given that do we need the statement manager?

**Comment by jhannus** [comment]
That would be great. When I read these lines I thought otherwise but I'm seeing other code where they are cached. https://github.com/datastax/ruby-driver/blob/v2.1.7/lib/cassandra/cluster/client.rb#L224-L236. I'll remove those commits.

**Comment by jhannus** [comment]
I rebased to remove those commits 👍

**Comment by pezra** [comment]
If you rebase i think this is ready to merge

**Comment by jhannus** [comment]
OK all rebased to master.

**Comment by pezra** [comment]
one more time. i merged another PR at the same time you did this.

**Comment by pezra** [comment]
Looks like a few things need to change due to the upgrade of the cassandra driver gem.

**Comment by jhannus** [comment]
OK, I got those changes in there as well. We'll see from the spec run if there are still issues with the upgrade.

**Comment by pezra** [comment]
Rebased because i merged another PR.

**Comment by pezra** [comment]
@jhannus, i rebased and, i think, fix the last remaining error. However that last rebase was a little weird. Could skim the current set of commits to ensure sure they make sense.

**Comment by jhannus** [comment]
Sure thing...

**Comment by jhannus** [comment]
Looks good 👍

**Comment by pezra** [comment]
merged! thanks for the improvement and the stick-to-it-iveness to get it to completion.

**Pull request merged**